From: Colin Watson <cjwatson@debian.org>
Date: Tue, 18 Mar 2025 16:05:59 +0000
Subject: Fix test failure in 2027

Santiago Vila pointed out in https://bugs.debian.org/1078638 that one of
the test cases in `tests/test_parser.py::test_parser` will start failing
soon, specifically at the start of 2027; this is because two-digit years
are converted to a year within [-50, 49] of the current year.

To avoid this, run tests of parsing two-digit years with a frozen time.

Forwarded: https://github.com/dateutil/dateutil/pull/1422
Bug-Debian: https://bugs.debian.org/1078638
Last-Update: 2025-03-18
---
 tests/test_parser.py | 33 +++++++++++++++++++++------------
 1 file changed, 21 insertions(+), 12 deletions(-)

diff --git a/tests/test_parser.py b/tests/test_parser.py
index ce9068a..cbd7a2d 100644
--- a/tests/test_parser.py
+++ b/tests/test_parser.py
@@ -12,6 +12,8 @@ from dateutil.parser import parse, parserinfo
 from dateutil.parser import ParserError
 from dateutil.parser import UnknownTimezoneWarning
 
+from freezegun import freeze_time
+
 from ._common import TZEnvContext
 
 from io import StringIO
@@ -36,6 +38,7 @@ def fuzzy(request):
 
 
 # Parser test cases using no keyword arguments. Format: (parsable_text, expected_datetime, assertion_message)
+# parsable_text may be a tuple of (parsable_text, frozen_time) to parse the given text at the given time.
 PARSER_TEST_CASES = [
     ("Thu Sep 25 10:36:28 2003", datetime(2003, 9, 25, 10, 36, 28), "date command format strip"),
     ("Thu Sep 25 2003", datetime(2003, 9, 25), "date command format strip"),
@@ -53,40 +56,41 @@ PARSER_TEST_CASES = [
     ("09-25-2003", datetime(2003, 9, 25), "date with dash"),
     ("25-09-2003", datetime(2003, 9, 25), "date with dash"),
     ("10-09-2003", datetime(2003, 10, 9), "date with dash"),
-    ("10-09-03", datetime(2003, 10, 9), "date with dash"),
+    (("10-09-03", datetime(2025, 1, 1)), datetime(2003, 10, 9), "date with dash"),
     ("2003.09.25", datetime(2003, 9, 25), "date with dot"),
     ("09.25.2003", datetime(2003, 9, 25), "date with dot"),
     ("25.09.2003", datetime(2003, 9, 25), "date with dot"),
     ("10.09.2003", datetime(2003, 10, 9), "date with dot"),
-    ("10.09.03", datetime(2003, 10, 9), "date with dot"),
+    (("10.09.03", datetime(2025, 1, 1)), datetime(2003, 10, 9), "date with dot"),
     ("2003/09/25", datetime(2003, 9, 25), "date with slash"),
     ("09/25/2003", datetime(2003, 9, 25), "date with slash"),
     ("25/09/2003", datetime(2003, 9, 25), "date with slash"),
     ("10/09/2003", datetime(2003, 10, 9), "date with slash"),
-    ("10/09/03", datetime(2003, 10, 9), "date with slash"),
+    (("10/09/03", datetime(2025, 1, 1)), datetime(2003, 10, 9), "date with slash"),
     ("2003 09 25", datetime(2003, 9, 25), "date with space"),
     ("09 25 2003", datetime(2003, 9, 25), "date with space"),
     ("25 09 2003", datetime(2003, 9, 25), "date with space"),
     ("10 09 2003", datetime(2003, 10, 9), "date with space"),
-    ("10 09 03", datetime(2003, 10, 9), "date with space"),
-    ("25 09 03", datetime(2003, 9, 25), "date with space"),
-    ("03 25 Sep", datetime(2003, 9, 25), "strangely ordered date"),
-    ("25 03 Sep", datetime(2025, 9, 3), "strangely ordered date"),
+    (("10 09 03", datetime(2025, 1, 1)), datetime(2003, 10, 9), "date with space"),
+    (("25 09 03", datetime(2025, 1, 1)), datetime(2003, 9, 25), "date with space"),
+    (("03 25 Sep", datetime(2025, 1, 1)), datetime(2003, 9, 25), "strangely ordered date"),
+    (("25 03 Sep", datetime(2025, 1, 1)), datetime(2025, 9, 3), "strangely ordered date"),
     ("  July   4 ,  1976   12:01:02   am  ", datetime(1976, 7, 4, 0, 1, 2), "extra space"),
-    ("Wed, July 10, '96", datetime(1996, 7, 10, 0, 0), "random format"),
+    (("Wed, July 10, '96", datetime(2025, 1, 1)), datetime(1996, 7, 10, 0, 0), "random format"),
     ("1996.July.10 AD 12:08 PM", datetime(1996, 7, 10, 12, 8), "random format"),
     ("July 4, 1976", datetime(1976, 7, 4), "random format"),
     ("7 4 1976", datetime(1976, 7, 4), "random format"),
     ("4 jul 1976", datetime(1976, 7, 4), "random format"),
     ("4 Jul 1976", datetime(1976, 7, 4), "'%-d %b %Y' format"),
-    ("7-4-76", datetime(1976, 7, 4), "random format"),
+    (("7-4-76", datetime(2026, 12, 31)), datetime(1976, 7, 4), "random format"),
+    (("7-4-76", datetime(2027, 1, 1)), datetime(2076, 7, 4), "random format"),
     ("19760704", datetime(1976, 7, 4), "random format"),
     ("0:01:02 on July 4, 1976", datetime(1976, 7, 4, 0, 1, 2), "random format"),
     ("July 4, 1976 12:01:02 am", datetime(1976, 7, 4, 0, 1, 2), "random format"),
     ("Mon Jan  2 04:24:27 1995", datetime(1995, 1, 2, 4, 24, 27), "random format"),
-    ("04.04.95 00:22", datetime(1995, 4, 4, 0, 22), "random format"),
+    (("04.04.95 00:22", datetime(2025, 1, 1)), datetime(1995, 4, 4, 0, 22), "random format"),
     ("Jan 1 1999 11:23:34.578", datetime(1999, 1, 1, 11, 23, 34, 578000), "random format"),
-    ("950404 122212", datetime(1995, 4, 4, 12, 22, 12), "random format"),
+    (("950404 122212", datetime(2025, 1, 1)), datetime(1995, 4, 4, 12, 22, 12), "random format"),
     ("3rd of May 2001", datetime(2001, 5, 3), "random format"),
     ("5th of March 2001", datetime(2001, 3, 5), "random format"),
     ("1st of May 2003", datetime(2003, 5, 1), "random format"),
@@ -106,7 +110,12 @@ assert len(set([x[0] for x in PARSER_TEST_CASES])) == len(PARSER_TEST_CASES)
 
 @pytest.mark.parametrize("parsable_text,expected_datetime,assertion_message", PARSER_TEST_CASES)
 def test_parser(parsable_text, expected_datetime, assertion_message):
-    assert parse(parsable_text) == expected_datetime, assertion_message
+    if isinstance(parsable_text, tuple):
+        parsable_text, frozen_time = parsable_text
+        with freeze_time(frozen_time):
+            assert parse(parsable_text, parserinfo=parserinfo()) == expected_datetime, assertion_message
+    else:
+        assert parse(parsable_text) == expected_datetime, assertion_message
 
 
 # Parser test cases using datetime(2003, 9, 25) as a default.
